src/Compilers/Shared/DesktopBuildClient.cs

```diff
@@ -313,6 +317,41 @@ protected override int RunLocalCompilation(string[] arguments, BuildPaths buildP
             }
         }
 
+        // Protected for testing
+        protected static bool TryConnectToNamedPipeWithSpinWait(NamedPipeClientStream pipeStream,
+                                                                int timeoutMs,
+                                                                CancellationToken cancellationToken)
+        {
+            // .NET 4.5 implementation of NamedPipeStream.Connect busy waits the entire time.
+            // Work around is to spin wait.
+            const int maxWaitInterval = 50;
+            int elapsed = 0;
+            var sw = new SpinWait();
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                int waitTime = Math.Min(timeoutMs - elapsed, maxWaitInterval);
+                if (timeoutMs != Timeout.Infinite && waitTime <= 0)
+                {
+                    return false;
+                }
+
+                try
+                {
+                    pipeStream.Connect(waitTime);
```

Comment:
Let's say timeoutMs == Timeout.Infinite, which is -1. We end up here with waitTime == -1. Do we want pipeStream.Connect to sit here, waiting indefinitely?